components/Ripple/src/MDCRippleTouchController.h

```diff
@@ -0,0 +1,184 @@
+ selected state where the ripple remains spread on the view.
+
+ - MDCRippleStateNormal: The ripple is in its normal state.
+ - MDCRippleStateSelected: The ripple is in the selected state.
+ */
+typedef NS_ENUM(NSInteger, MDCRippleState) {
+  MDCRippleStateNormal = 0,
+  MDCRippleStateSelected,
+};
+
+/**
+ The MDCRippleTouchController is a convenience controller that adds all the needed touch tracking
+ and logic to provide the correct ripple effect based on the user interacting with the view the
+ ripple is added to.
+ */
+@interface MDCRippleTouchController : NSObject <UIGestureRecognizerDelegate>
+
+/**
+ A weak reference to the view that the ripple is added to, and that responds to the user
+ events.
+ */
+@property(nonatomic, weak, readonly, nullable) UIView *view;
+
+/**
+ The ripple view added to the view.
+ */
+@property(nonatomic, strong, readonly, nonnull) MDCRippleView *rippleView;
+
+/** Delegate to extend the behavior of the touch control. */
+
+/**
+ A delegate to extend the behavior of the touch controller.
+ */
+@property(nonatomic, weak, nullable) id<MDCRippleTouchControllerDelegate> delegate;
+
+/** Gesture recognizer used to bind touch events to ripple. */
+
+/**
+ The gesture recognizer used to bind the touch events to the ripple.
+ */
+@property(nonatomic, strong, readonly, nonnull) UILongPressGestureRecognizer *gestureRecognizer;
+
+/**
+ The selection gesture recognizer used to bind the touch events related to selection to the ripple.
+ */
+@property(nonatomic, strong, readonly, nullable)
+    UILongPressGestureRecognizer *selectionGestureRecognizer;
+
+/**
+ This BOOL tells the touch controller if to allow selection and all the logic and visuals
+ that come with it, for this ripple.
+
+ Defaults to NO.
+ */
+@property(nonatomic) BOOL allowsSelection;
+
+/**
+ This BOOL is set to YES if the ripple is currently selected, or NO otherwise.
+ It only has significance if selectionMode is activated.
+
+ Defaults to NO.
+ */
+@property(nonatomic, getter=isSelected) BOOL selected;
+
+/**
+ This BOOL is set to YES if the ripple is currently in the selection mode, or NO otherwise.
+
+ Defaults to NO.
+ */
+@property(nonatomic) BOOL selectionMode;
+
+/**
+ Sets the color of the ripple for state.
+
+ @param rippleColor The ripple color to set the ripple to.
+ @param state The state of the ripple in which to set the ripple color.
+ */
+- (void)setRippleColor:(nullable UIColor *)rippleColor forState:(MDCRippleState)state;
+
+/**
+ Gets the ripple color for the given state.
+
+ @param state The ripple's state.
+ @return the color of the ripple for state.
+ */
+- (nullable UIColor *)rippleColorForState:(MDCRippleState)state;
+
+/**
+ Sets the alpha of the ripple for state.
+
+ @param rippleAlpha The ripple alpha to set the ripple to.
+ @param state The state of the ripple in which to set the ripple alpha.
+ */
+- (void)setRippleAlpha:(CGFloat)rippleAlpha forState:(MDCRippleState)state;
```

Comment:
If a client sets the color to be [UIColor.blue withAlphaComponent:0.5] while using `setRippleColor:ForState:` and then sets the alpha to `0.5` for `setRippleAlpha:forState` will the alpha be multiplied? I think we can remove this.

Reply:
The ripple's alpha needs to be different when it retains on the screen when being selected or not. the ripple's color needs to be different when moving in and out of selection mode. the separation allows us to have that fine amount of customization. an alternative is to provide another state system and have a matrix of colors between the 2 state systems, because there can be cases where there is a selected color with a normal alpha. I will leave this like this for now, and polish it once i work more closely with design on the selected state.